redfish/resource_redfish_bios.go

```diff
@@ -0,0 +1,268 @@
+				if intVal != intOldVal {
+					attrsPayload[key] = intVal
+				}
+			} else {
+				if val != oldVal {
+					attrsPayload[key] = val
+				}
+			}
+
+		} else {
+			return diag.Errorf("BIOS attibute %s not found", key)
+		}
+	}
+
+	if len(attrsPayload) != 0 {
+		if !pending {
+			err = updateBiosAttributes(d, bios, attrsPayload)
+			if err != nil {
+				return diag.Errorf("error updating bios attributes: %s", err)
+			}
+		}
+	} else {
+		log.Printf("[DEBUG] BIOS attributes are already set")
+	}
+
+	if err := d.Set("attributes", attributes); err != nil {
+		return diag.Errorf("error setting bios attributes: %s", err)
+	}
+
+	// Set the ID to the @odata.id
+	d.SetId(bios.ODataID)
+
+	log.Printf("[DEBUG] %s: Update finished successfully", d.Id())
+	return diags
+}
+
+func resourceRedfishBiosRead(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+
+	log.Printf("[DEBUG] %s: Beginning read", d.Id())
+	var diags diag.Diagnostics
+
+	conn := m.(*gofish.APIClient)
+
+	bios, err := getBios(conn)
+	if err != nil {
+		return diag.Errorf("error fetching BIOS resource: %s", err)
+	}
+
+	attributes := make(map[string]string)
+	err = copyBiosAttributes(bios, attributes)
+	if err != nil {
+		return diag.Errorf("error fetching BIOS attributes: %s", err)
+	}
+
+	if err := d.Set("attributes", attributes); err != nil {
+		return diag.Errorf("error setting bios attributes: %s", err)
+	}
+
+	// Set the ID to the @odata.id
+	d.SetId(bios.ODataID)
+
+	log.Printf("[DEBUG] %s: Read finished successfully", d.Id())
+
+	return diags
+}
+
+func resourceRedfishBiosDelete(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+
+	var diags diag.Diagnostics
+
+	d.SetId("")
+
+	return diags
+}
+
+func getBios(conn *gofish.APIClient) (*redfish.Bios, error) {
+
+	service := conn.Service
+	systems, err := service.Systems()
+	if err != nil {
+		return nil, err
+	}
+
+	bios, err := systems[0].Bios()
+	if err != nil {
+		return nil, err
+	}
+
+	return bios, nil
+}
+
+func copyBiosAttributes(bios *redfish.Bios, attributes map[string]string) error {
+
+	// TODO: BIOS Attributes' values might be any of several types.
+	// terraform-sdk currently does not support a map with different
+	// value types. So we will convert int and float values to string.
+	// copy from the BIOS attributes to the new bios attributes map
+	for key, value := range bios.Attributes {
+		if attr_val, ok := value.(string); ok {
+			attributes[key] = attr_val
+		} else {
+			attributes[key] = fmt.Sprintf("%v", value)
+		}
+	}
+
+	return nil
+}
+
+func updateBiosAttributes(d *schema.ResourceData, bios *redfish.Bios, attributes map[string]interface{}) error {
+
+	payload := make(map[string]interface{})
+	payload["Attributes"] = attributes
+
+	if settingsApplyTime, ok := d.GetOk("settings_apply_time"); ok {
+		allowedValues := bios.AllowedAttributeUpdateApplyTimes()
+		allowed := false
+		for i := range allowedValues {
```

Comment:
Hi Anupam,

If we are using *validation.StringInSlice* in **ValidateFunc** in the schema to check if the value given on **settings_apply_time** is within four values, why do we check that again manually?

Reply:
@mikeletux, the schema validation check for ```settings_apply_time``` checks it against all possible values as per the spec, however, a BMC might only support a subset of those. That is why I added the check to make sure that the value given is being supported by the BMC or not.